### PR TITLE
[RISCV] RISC-V Mimiker setup

### PR DESCRIPTION
**Pull request by MichalBlk**
Introduce an initial setup of the RISC-V Mimiker port.

### CODE REVIEW

File: toolchain/gnu/Makefile [hunk 1]
```diff
@@ -166,7 +166,7 @@ $(SRCDIR)/$(ISL).tar.xz:
 
 isl-unpack: $(SRCDIR)/isl/.unpack
 $(SRCDIR)/isl/.unpack: $(SRCDIR)/$(ISL).tar.xz
-	$(CD) $(SRCDIR) && $(TAR) xJf $^ && $(MV)  $(ISL) isl
+	$(CD) $(SRCDIR) && $(TAR) xf $^ && $(MV)  $(ISL) isl
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply toolchain/gnu/Makefile]: It has already been fixed by @cahirwpz in #1179 .

File: toolchain/gnu/config.mk [hunk 2]
```diff
@@ -14,7 +14,7 @@ BINUTILS = binutils-2.37
 GCC = gcc-11.2.0
 GDB = gdb-11.1
 
-ISL-URL = "http://isl.gforge.inria.fr/$(ISL).tar.xz"
+ISL-URL = "https://gcc.gnu.org/pub/gcc/infrastructure/$(ISL).tar.bz2"
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply toolchain/gnu/config.mk]: Same as above (#1179 ).

File: lib/csu/riscv/ld.script [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  {
+    *(.bss .bss.*)
+    *(.common)
+    . = ALIGN(4);
+    HIDDEN(__ebss = .);
+  } : bss
+
+  HIDDEN(_gp_off = 4K / 2);
+  __global_pointer$ = MIN(__sdata + _gp_off,
+                          MAX(__data + _gp_off, __ebss - _gp_off));
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply lib/csu/riscv/ld.script:66]: NetBSD also packs `.sbss` and `.bss` in a single `.bss` output section (http://bxr.su/NetBSD/sys/arch/riscv/conf/kern.ldscript).

File: launch [hunk 4]
```diff
@@ -103,7 +109,20 @@ CONFIG = {
                     '-cpu', 'cortex-a53'],
                 'network_options': [],
                 'uarts': [
-                    dict(name='/dev/cons', port=RandomPort(), raw=True)
+                    dict(name='/dev/tty0', port=RandomPort(), raw=True)
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply launch:112]: There is no relation. A separate PR will be made.

File: launch [hunk 5]
```diff
@@ -320,12 +370,12 @@ def DevelRun(sim, dbg):
             dbg.start(session)
 
         session.kill_window(':0')
-        session.select_window(dbg.name if dbg else '/dev/tty1')
+        session.select_window(dbg.name if dbg else '/dev/tty0')
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply launch:373]: It's purely a cosmetic change. Ttys are usually enumerated from 0.

File: launch [hunk 6]
```diff
@@ -89,8 +95,8 @@ CONFIG = {
                     '-object', 'filter-dump,id=net0,netdev=net0,file=rtl.pcap'
                 ],
                 'uarts': [
-                    dict(name='/dev/tty1', port=RandomPort(), raw=True),
-                    dict(name='/dev/tty2', port=RandomPort()),
+                    dict(name='/dev/tty0', port=RandomPort(), raw=True),
+                    dict(name='/dev/tty1', port=RandomPort()),
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply launch:99]: Same as below.

File: include/riscv/mcontext.h [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+#define _UC_SETSTACK 0x00010000 /* see <sys/ucontext.h> */
+#define _UC_CLRSTACK 0x00020000 /* see <sys/ucontext.h> */
+#define _UC_TLSBASE 0x00080000  /* see <sys/ucontext.h> */
+
+#define _UC_MACHINE_SP(uc) ((uc)->uc_mcontext.__gregs[_REG_SP])
+#define _UC_MACHINE_FP(uc) ((uc)->uc_mcontext.__gregs[_REG_S0])
+#define _UC_MACHINE_PC(uc) ((uc)->uc_mcontext.__gregs[_REG_PC])
+#define _UC_MACHINE_INTRV(uc) ((uc)->uc_mcontext.__gregs[_REG_RV])
+
+#define _UC_MACHINE_SET_PC(uc, pc) _UC_MACHINE_PC(uc) = (pc)
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply include/riscv/mcontext.h:121]: I agree.

File: include/riscv/asm.h [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+#else
+#define PLT(x) x
+#endif
+
+/*
+ * WEAK_ALIAS: create a weak alias.
+ */
+#define WEAK_ALIAS(alias, sym)                                                 \
+  .weak alias;                                                                 \
+  alias = sym
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply include/riscv/asm.h]: Alright. It's been removed.

File: build/tools.mk [hunk 9]
```diff
@@ -33,6 +33,7 @@ endif
 CC       = $(TARGET_CC)
 AS       = $(TARGET_CC)
 LD       = $(TARGET)-gcc $(GCC_ABIFLAGS) -g
+CPP      = $(TARGET)-cpp -E
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply build/tools.mk]: I've changed `ccp -E' to `gcc -E`.

**Comment by MichalBlk** [comment]
I will divide the PR into a few sub PRs as @pj1031999 has suggested.

**Comment by MichalBlk** [comment]
This PR has been moved to "blocked" and waits for PRs #1183, #1184, #1185, #1186.

File: lib/csu/riscv/ld.script (hunk 3 above)
[the remark this answers is not in the document]
Comment (cahirwpz) [reply lib/csu/riscv/ld.script:66]: @MichalBlk I don't think we should take MIPS as an example. We had a few problems with `.sdata` and `.sbss` and finally I added `-G 0` option to get rid of relative addressing.

File: sys/riscv/riscv.ld.in [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    *(.sbss .scommon)
+    *(.bss .bss.*)
+    *(COMMON)
+    __ebss = ABSOLUTE(.);
+    __kernel_end = ABSOLUTE(.);
+  }
+
+  /* Provide global pointer for linker relaxations. */
+  HIDDEN(_gp_off = 4K / 2);
+  __global_pointer$ = MIN(__sdata + _gp_off,
```
Comment (cahirwpz) [inline sys/riscv/riscv.ld.in:65]: Ditto. We need to discuss it.

File: sys/riscv/start.S [hunk 11]
```diff
@@ -0,0 +1,10 @@
+#include <riscv/asm.h>
+
+	.org 0x0
```
Comment (cahirwpz) [inline sys/riscv/start.S:3]: Is it really loaded at address 0 ?

File: sys/tests/vm_map.c [hunk 12]
```diff
@@ -10,11 +10,9 @@
 #include <sys/sched.h>
 #include <sys/proc.h>
 
-#ifdef __mips__
+#if __SIZEOF_POINTER__ == 4
```
Comment (cahirwpz) [inline sys/tests/vm_map.c:13]: Ditto. `_LP64`

**Review by cahirwpz — COMMENTED** [review]
Please address my comments.

File: sys/riscv/start.S (hunk 11 above)
Comment (MichalBlk-I) [reply sys/riscv/start.S]: Done.

File: include/riscv/asm.h [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+ */
+
+/* MODIFIED */
+
+#ifndef _RISCV_ASM_H
+#define _RISCV_ASM_H
+
+#define _C_LABEL(x) x
+
+#define __CONCAT(x, y) x##y
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply include/riscv/asm.h]: Done.

File: include/riscv/mcontext.h [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+
+/* MODIFIED */
+
+#ifndef _RISCV_MCONTEXT_H_
+#define _RISCV_MCONTEXT_H_
+
+#include <sys/types.h>
+
+#if defined(__riscv_f) || defined(__riscv_d)
+#define FPE 1
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply include/riscv/mcontext.h]: Done.

File: include/riscv/cdefs.h [hunk 15]
```diff
@@ -0,0 +1,6 @@
+#ifndef _RISCV_CDEFS_H_
+#define _RISCV_CDEFS_H_
+
+#define __ALIGNBYTES (sizeof(long long) - 1)
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply include/riscv/cdefs.h]: This has been changed in #1201.

File: lib/csu/riscv/ld.script [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+
+  . = ALIGN(4096);
+  .rodata :
+  {
+    *(.rodata .rodata.*)
+    . = ALIGN(4);
+  } : rodata
+
+  . = ALIGN(4096);
+  .data :
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply lib/csu/riscv/ld.script:32]: Right. This has been changed and simplified.

File: lib/csu/riscv/ld.script [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+  . = ALIGN(4096);
+  .sdata :
+  {
+    HIDDEN(__sdata = .);
+    *(.sdata .sdata.*)
+    . = ALIGN(4);
+  } : sdata
+
+  . = ALIGN(4096);
+  .sbss :
```
[the remark this answers is not in the document]
Comment (MichalBlk) [reply lib/csu/riscv/ld.script:50]: I agree, done.

**Review by cahirwpz — APPROVED** [review]
Ok. I feel pretty confident about _linker scripts_. I don't fully understand why linker relaxations are used only for `.sdata`, nonetheless linker should provide well-formed executables.

**Pull request merged**